File: puzzles/haskell/dont-panic1.hs
import System.IO
import Control.Monad
import Data.Maybe

main :: IO ()
main = do
  hSetBuffering stdout NoBuffering -- DO NOT REMOVE
  
  input_line <- getLine
  let input = words input_line
  let exitFloor = read (input!!3) :: Int -- floor on which the exit is found
  let exitPos = read (input!!4) :: Int -- position of the exit on its floor
  let nbElevators = read (input!!7) :: Int -- number of elevators
  
  elevators <- replicateM nbElevators $ do
    input_line <- getLine
    let input = words input_line
    let elevatorFloor = read (input!!0) :: Int -- floor on which this elevator is found
    let elevatorPos = read (input!!1) :: Int -- position of the elevator on its floor
    return (elevatorFloor, elevatorPos)
  
  -- game loop
  forever $ do
    input_line <- getLine
    let input = words input_line
    let cloneFloor = read (input !! 0) :: Int -- floor of the leading clone
    let clonePos = read (input !! 1) :: Int -- position of the leading clone on its floor
    let direction = input !! 2 -- direction of the leading clone: LEFT or RIGHT
    
    if cloneFloor == -1 then
      putStrLn "WAIT"
    else if cloneFloor == exitFloor then
      if direction == "LEFT" && clonePos < exitPos then
        putStrLn "BLOCK"
      else if direction == "RIGHT" && clonePos > exitPos then
        putStrLn "BLOCK"
      else
        putStrLn "WAIT"
    else do
      let elevatorPos = fromJust (lookup cloneFloor elevators)
      if direction == "LEFT" && clonePos < elevatorPos then
        putStrLn "BLOCK"
      else if direction == "RIGHT" && clonePos > elevatorPos then
        putStrLn "BLOCK"
      else
        putStrLn "WAIT"

File: puzzles/python3/dont_panic1.py
if __name__ == "__main__":
    elevators = {}

    _, _, _, exit_floor, exit_pos, _, _, n5 = map(int, input().split())
    for _ in range(n5):
        elevator_floor, elevator_pos = map(int, input().split())
        elevators[elevator_floor] = elevator_pos
    # For the last floor, we use the exit instead of an elevator.
    elevators[exit_floor] = exit_pos

    # game loop
    while True:
        clone_floor, clone_pos, direction = input().split()
        clone_floor = int(clone_floor)
        clone_pos = int(clone_pos)
        if clone_floor == -1:
            print("WAIT")
        elif direction == "LEFT" and clone_pos < elevators[clone_floor]:
            print("BLOCK")
        elif direction == "RIGHT" and clone_pos > elevators[clone_floor]:
            print("BLOCK")
        else:
            print("WAIT")
